### ISSUE TITLE
Cannot link Glulxe against libgarglk

### ISSUE THREAD

**Issue by curiousdannii** [garglk/garglk#482]
So I'm trying to link glulxe against libgarglk, and there are lots of errors.

Firstly, I'm building garglk like this (from within the glulxe folder):

```
curl -Ls https://github.com/garglk/garglk/archive/refs/heads/master.tar.gz | tar xz
mv garglk-master garglk -f
mkdir garglk/build
cmake garglk -B garglk/build \
    -DBUILD_SHARED_LIBS=OFF -DCMAKE_BUILD_TYPE=Release \
    -DWITH_BABEL=OFF -DWITH_INTERPRETERS=OFF -DWITH_LAUNCHER=OFF
make -C garglk/build -j$(nproc) --no-print-directory
cp garglk/garglk/*.h garglk/build/garglk
```

I then modified glulxe's Makefile to have this at the top

```
GLKINCLUDEDIR = ./garglk/build/garglk
GLKLIBDIR = ./garglk/build/garglk
#GLKMAKEFILE = Make.cheapglk
GLKLIB = -lgarglk
```

And comment out the `include $(GLKINCLUDEDIR)/$(GLKMAKEFILE)` line. (It would be nice if garglk make a Make.garglk file, but that's a very minor issue.)

When running make, it then puts out hundreds of errors, saying that it has undefined references to all the `FT_` and `Q...` functions. It also doesn't have `main`! I noticed this is because it only includes main.c when `WITH_INTERPRETERS` is on.

So I don't know if it would only take a few small changes to make it so that terps like Glulxe can be directly linked against libgarglk, or if it will require really massive ones. I don't think I'll be much use for fixing any of this, but I'd be happy to test anything you come up with.

**Comment by curiousdannii** [comment]
Actually, seeing how xglk does add the rest of the libs to LINKLIBS, I think making a Make.garglk file might be the solution.

https://github.com/erkyrath/xglk/blob/master/Makefile

I've only ever used cheapglk and remglk before, which don't have any extra libs, so their Make.x files aren't particularly useful.

**Comment by cspiegel** [comment]
The root of the problem is that `libgarglkmain` needs to be linked in as well.  That's a bit of glue code that's necessary for a shared `libgarglk`: the shared library can't include `main` because that calls `glk_main`, which won't exist till an interpreter is actually linked in.  `libgarglkmain` is a static library, which _can_ easily include references to non-existent symbols (since it's just an archive of object files).  So you have `main` in `libgarglkmain` which calls `glk_main`, provided by the interpreter, which then makes calls to Glk functions in `libgarglk`.

I have this halfway done in Bocfel, using exactly the method you suggest: I provide a `Make.gargoyle` that looks like this:

```
GLKLIB = -lgarglk -lgarglkmain
```
But that's not really compatible with how glulxe expects things to be laid out.  Glulxe wants `GLKINCLUDEDIR` to contain both the headers and the `Make.gargoyle` (or whatever it's called), but that's not the way Unix packages tend to be installed.  In short, Glulxe expects you to build a Glk library in its own directory and then point to that directory.  It doesn't expect a system-wide Glk installation.

That's not an insurmountable problem.  I can set up the installer to create a directory like `/usr/lib/gargoyle/build` to contain `Make.gargoyle` and put the headers in there.  It would set `GLKLIB` to `-L[PREFIX]/lib -lgarglk -lgarglkmain`.

Right now Gargoyle installes a `.pc` file for `pkg-config` which is the modern Unix way of handling libraries, but since no other Glk implementation does that, it's not necessarily useful.  Prior to the CMake system, `libgarglk` wasn't meant to be used outside of Gargoyle, so I'm still in a "testing" phase, so to speak, of how to get it working right, and your feedback is really helpful.

However, I'm not sure about why you're getting undefined references to Qt/FreeType functions.  Those ought to be linked in via the shared library (i.e. it should link to all its necessary libraries, meaning there won't be any undefined references).  As a quick test, can you include the output of `ldd /path/to/libgarglk.so`?  I'd be interested in which libraries it's linking to...

> Another thought: maybe there could be a new build option for this stand alone Glk library pattern? It would include main.c and create Make.garglk, as well as setting it to be a static library. It might be better to do this as a separate build option rather than overloading BUILD_SHARED_LIBS, which I assume has its own uses within the Gargoyle project.

I saw this addition after I wrote the above.  First, `BUILD_SHARED_LIBS` is really only used to build a shared `libgarglk.so`, so it'd be fine, I think, to use that to determine behavior for static vs shared.  But I'm not 100% sure about putting together a static library.  At the very least it would need to include references to _all_ other shared libraries (Qt, FreeType, fontconfig, SDL, etc).  Not that that's impossible, but may be a pain to tease out.  And it's complicated by C++: `libstdc++` would need to be linked in as well, except not all platforms use `libstdc++`, and I don't know if CMake specifically calls out the C++ library, because it's linking with the C++ compiler which will automatically pull in the required library; so Gargoyle could just require linking with C++, but that's not particularly friendly either...

In short, we'll get the thing solved but it'll probably take a few tries. I'll start cobbling some stuff together that you can look at.

Thank you for the report!

**Comment by curiousdannii** [comment]
> The root of the problem is that libgarglkmain needs to be linked in as well. That's a bit of glue code that's necessary for a shared libgarglk: the shared library can't include main because that calls glk_main, which won't exist till an interpreter is actually linked in. libgarglkmain is a static library, which can easily include references to non-existent symbols (since it's just an archive of object files). So you have main in libgarglkmain which calls glk_main, provided by the interpreter, which then makes calls to Glk functions in libgarglk.

Ah, that makes sense. I hadn't thought about it, but of course when libgarglk is a shared library the normal (for Glk, but very weird for other libraries) pattern of having the library contain `main` wouldn't work.

libgarglkmain is currently gated behind WITH_INTERPRETERS, which I didn't want to turn on, mostly because it would waste time when I only want libgarglk. Does it need to be gated behind WITH_INTERPRETERS? Can it just always be built?

> In short, Glulxe expects you to build a Glk library in its own directory and then point to that directory. It doesn't expect a system-wide Glk installation.
>
> That's not an insurmountable problem. I can set up the installer to create a directory like /usr/lib/gargoyle/build to contain Make.gargoyle and put the headers in there. It would set GLKLIB to -L[PREFIX]/lib -lgarglk -lgarglkmain.

I hadn't been intending to install it at all. I've been mostly thinking of this for CI scripts, or for developers using [Glulxtoc](https://github.com/curiousdannii/if-decompiler), who might want to build against garglk. (That's what got me started here, I want to document in full how to build against garglk with a Glulxtoc project. But it's easiest to start with Glulxe.) I think it would be important to be able to do all of this without installing garglk - we don't want to clobber whatever proper Gargoyle installation they currently have.

> However, I'm not sure about why you're getting undefined references to Qt/FreeType functions. Those ought to be linked in via the shared library (i.e. it should link to all its necessary libraries, meaning there won't be any undefined references). As a quick test, can you include the output of ldd /path/to/libgarglk.so? I'd be interested in which libraries it's linking to...

I'm not using a share library, I'm using building a static library. Mostly because that's what I'm used to with other Glk libraries, and also because the idea of getting a single binary at the end is appealing. I haven't tried it with a shared library, I can, but not right now.

>In short, we'll get the thing solved but it'll probably take a few tries. I'll start cobbling some stuff together that you can look at.
>
>Thank you for the report!

Thank you so much!

**Comment by cspiegel** [comment]
> libgarglkmain is currently gated behind WITH_INTERPRETERS, which I didn't want to turn on, mostly because it would waste build time when I only want libgarglk. Does it need to be gated behind WITH_INTERPRETERS? Can it just always be built?

It can and should always be built.  Total mistake on my part to make it conditional on the interpreters.  It's a small part of this issue, but an easy fix, so I'll get that in first.

**Comment by cspiegel** [comment]
I've got a tentative approach to this in the branch https://github.com/cspiegel/garglk/tree/glk-layout

If you pass `-DGLK_LAYOUT=ON` to CMake, it'll disable the interpreters and launcher.  You still have to pass `-DBUILD_SHARED_LIBS=OFF` if you want a static library.

After building, `make install` will put things into `build/dist` in the top level source directory.  This contains `Make.garglk` as well as headers and libraries.  `Make.garglk` will, if the library is static, include all (hopefully) library dependencies required to link.  I tested by setting variables in glulxe as follows:

```
GLKINCLUDEDIR = /home/chris/github/garglk/build/dist
GLKLIBDIR = /home/chris/github/garglk/build/dist
GLKMAKEFILE = Make.garglk
```
There is one catch: things will still fail to link due to a missing C++ library.  As I briefly mentioned earlier, this isn't a very easy problem to solve, as far as I can see.  You can manually link in the standard C++ library for your platform, but I can't determine what that is in CMake.  I'm not sure it's _possible_ to do that in any non-horrible way, given the fact that even knowing the compiler doesn't tell you which C++ library is being used (Clang can easily use either `libstdc++` or `libc++`, for example).

The other option is to link with a C++ compiler.

Either way it's a one-line change in glulxe (and probably other interpreters).  Hopefully this gets you a bit closer to your goal.

**Comment by cspiegel** [comment]
@curiousdannii Have you had a chance to look at this?  If it works, I'll get it merged in, but if there's more that can be done, let me know, and I'll see what I can do.

**Comment by curiousdannii** [comment]
Sorry, I haven't had time recently. Hopefully soon!
